sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c

```diff
@@ -0,0 +1,1546 @@
+        gnrc_pktbuf_release(pkt);
+    }
+}
+
+int gnrc_sixlowpan_frag_sfr_forward(gnrc_pktsnip_t *pkt,
+                                    sixlowpan_sfr_rfrag_t *rfrag,
+                                    gnrc_sixlowpan_frag_vrb_t *vrbe,
+                                    unsigned page)
+{
+    _generic_rb_entry_t entry = { .type = _VRB, .entry = { .vrb = vrbe } };
+    gnrc_pktsnip_t *hdrsnip = gnrc_pktbuf_add(pkt, rfrag, sizeof(*rfrag),
+                                              GNRC_NETTYPE_SIXLOWPAN);
+
+    /* free all intervals associated to the VRB entry, as we don't need them
+     * with SFR, so throw them out, to save this resource */
+    while (vrbe->super.ints) {
+        vrbe->super.ints->end = 0U;
+        vrbe->super.ints = vrbe->super.ints->next;
+    }
+    if (hdrsnip == NULL) {
+        DEBUG("6lo sfr: Unable to allocate new rfrag header\n");
+        gnrc_pktbuf_release(pkt);
+        return -ENOMEM;
+    }
+    DEBUG("6lo sfr: adapting old fragment size (%u) for forwarding to %u\n",
+          sixlowpan_sfr_rfrag_get_frag_size(hdrsnip->data),
+          gnrc_pkt_len(pkt));
+    /* due to compression, packet length of the original fragment might have
+     * changed */
+    sixlowpan_sfr_rfrag_set_frag_size(hdrsnip->data, gnrc_pkt_len(pkt));
+    /* offset is adapted in `_forward_rfrag()` */
+    return _forward_rfrag(hdrsnip, &entry, sixlowpan_sfr_rfrag_get_offset(rfrag),
+                          page);
+}
+
+void gnrc_sixlowpan_frag_sfr_arq_timeout(gnrc_sixlowpan_frag_fb_t *fbuf)
+{
+    uint32_t now = xtimer_now_usec();
+    _frag_desc_t * const head = (_frag_desc_t *)fbuf->sfr.window.next;
+    _frag_desc_t *frag_desc = head;
+    uint32_t next_arq_offset = fbuf->sfr.arq_timeout;
+    bool reschedule_arq_timeout = false;
+
+    DEBUG("6lo sfr: ARQ timeout for datagram %u\n", fbuf->tag);
+    fbuf->sfr.arq_timeout_event.msg.content.ptr = NULL;
+    /* copying clist_foreach because we can't work just in function context*/
+    if (frag_desc) {
+        do {
+            uint32_t diff;
+
+            frag_desc = (_frag_desc_t *)frag_desc->super.next;
+            diff = now - frag_desc->last_sent;
+            if (diff < (fbuf->sfr.arq_timeout * US_PER_MS)) {
+                /* this fragment's last was last sent < fbuf->sfr.arq_timeout
+                 * ago */
+                uint32_t offset = fbuf->sfr.arq_timeout - (diff / US_PER_MS);
+
+                DEBUG("6lo sfr: wait for fragment %u in next reschedule\n",
+                      _frag_seq(frag_desc));
+                if (offset < next_arq_offset) {
+                    /* wait for this fragments ACK next */
+                    next_arq_offset = offset;
+                    DEBUG("         (next ARQ timeout in %lu)\n",
+                          (long unsigned)next_arq_offset);
+                }
+                reschedule_arq_timeout = true;
+            }
+            else if (_frag_ack_req(frag_desc)) {
```

Comment:
Would be good to have a line to explain what those different cases are.

Reply:
Done.